### PR TITLE
Use a more specific template object to remove docker-plugin dependency

### PR DESCRIPTION
**Pull request by ndeloof**

### CODE REVIEW

**Comment by ndeloof** [comment]
@reviewbybees

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    private final String remoteFs;
+
+    private final int instanceCap;
+
+    private final String label;
+
+
+    @DataBoundConstructor
+    public PodTemplate(String name, String image, boolean privileged, String command, String remoteFs, int instanceCap, String label) {
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: Recommend `@DataBoundSetter` for optional parameters.

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/PodTemplate/help-image.html [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Docker image ID for a jenkins JNLP slave.
+This image is responsible to run a jenkins jnlp bootsrap agent and connect to Jenkins master.
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/PodTemplate/help-image.html]: typo

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/PodTemplate/config.jelly [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  <f:entry field="label" title="${%Labels}">
+    <f:textbox/>
+  </f:entry>
+
+  <f:entry field="image" title="${%Docker image}">
+    <f:textbox/>
+  </f:entry>
+
+  <f:entry field="remoteFs" title="${%Remote FileSystem root}">
```
Comment (jglick) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/PodTemplate/config.jelly]: Meaningless title from a user perspective, rephrase.

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 4]
```diff
@@ -486,7 +472,7 @@ public ListBoxModel doFillCredentialsIdItems(@QueryParameter URL serverUrl) {
 
     @Override
     public String toString() {
-        return MoreObjects.toStringHelper(this).add("name", name).add("serverUrl", serverUrl).toString();
+        return "KubernetesCloud name:" + name +" serverUrl:" + serverUrl;
```
Comment (tfennelly) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: NIT: could have been replaced with `String.format`?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java [hunk 5]
```diff
@@ -39,25 +32,23 @@
     private final KubernetesCloud cloud;
 
     @DataBoundConstructor
-    public KubernetesSlave(DockerTemplate template, String nodeDescription, KubernetesCloud cloud, Label label)
+    public KubernetesSlave(PodTemplate template, String nodeDescription, KubernetesCloud cloud, Label label)
```
Comment (tfennelly) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java]: I presume backward compatibility is not a concern?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import hudson.model.Label;
+import hudson.model.labels.LabelAtom;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.util.Set;
+
+/**
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+public class PodTemplate extends AbstractDescribableImpl<PodTemplate> {
```
Comment (tfennelly) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: NIT: How about a useful class descripition?

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/config.jelly [hunk 7]
```diff
@@ -45,8 +45,8 @@
     </f:entry>
 
     <f:entry title="${%Images}" description="${%List of Images to be launched as slaves}">
-      <f:repeatableHeteroProperty field="templates" hasHeader="true" addCaption="Add Docker Template"
-                                  deleteCaption="Delete Docker Template" />
+      <f:repeatableHeteroProperty field="templates" hasHeader="true" addCaption="Add Pod Template"
+                                  deleteCaption="Delete Template" />
```
Comment (tfennelly) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud/config.jelly]: NIT: Wouldn't "Delete Pod Template` be more consistent with "Add Pod Template"?

**Comment by tfennelly** [comment]
:bee:

**Comment by recena** [comment]
@ndeloof Is not there a Jira ticket for that?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java (hunk 1 above)
Comment (recena) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplate.java]: +1

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java (hunk 5 above)
Comment (recena) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesSlave.java]: @tfennelly The plugin has only two release, it is a *young plugin*. If we are very strict with that, in one year we'll have more deprecated code than functional code :)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java (hunk 4 above)
Comment (ndeloof) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java]: String.format is java 1.7

File: src/main/resources/org/csanchez/jenkins/plugins/kubernetes/PodTemplate/config.jelly (hunk 3 above)
Comment (ndeloof) [inline src/main/resources/org/csanchez/jenkins/plugins/kubernetes/PodTemplate/config.jelly]: same for https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/hudson/slaves/DumbSlave/configure-entries.jelly#L39-39

**Pull request merged**